File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>

using namespace std;

bool has_matching_parentheses(string s) {
  vector<char> stack;

  for (char& ch : s) {
    if (ch == '(' || ch == '{' || ch == '[')
      stack.push_back(ch);
    else {
      if (stack.empty()) return false;

      int top = stack.back();

      if (ch == ')' && top == '(')
        stack.pop_back();
      else if (ch == '}' && top == '{')
        stack.pop_back();
      else if (ch == ']' && top == '[')
        stack.pop_back();
      else
        return false;
    }
  }
  return stack.empty();
}

int solution(string s) {
  int answer = 0;
  int len = s.length();

  for (int i = 0; i < len; i++) {
    if (has_matching_parentheses(s.substr(i) + s.substr(0, i))) ++answer;
  }
  return answer;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.Stack;

public class Solution {

    private static boolean hasMatchingParentheses(String s) {
        Stack<Character> stack = new Stack<>();

        for (char ch : s.toCharArray()) {
            if (ch == '(' || ch == '{' || ch == '[') {
                stack.push(ch);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }

                char top = stack.peek();
                if (ch == ')' && top == '(') {
                    stack.pop();
                } else if (ch == '}' && top == '{') {
                    stack.pop();
                } else if (ch == ']' && top == '[') {
                    stack.pop();
                } else {
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }

    public int solution(String s) {
        int answer = 0;
        int len = s.length();

        for (int i = 0; i < len; i++) {
            if (hasMatchingParentheses(s.substring(i) + s.substring(0, i))) {
                ++answer;
            }
        }
        return answer;
    }
}